src/test/scala/com/dwolla/mysql/init/RoundTripIntegrationTest.scala

```diff
@@ -0,0 +1,212 @@
+   * Secrets Manager service. That said, if you want to run the test that way, replace the
+   * {{{FakeSecretsManagerAlg}}} with {{{SecretsManagerAlg.resource[IO]}}}.
+   */
+  private val secretsManagerAlg: Fixture[SecretsManagerAlg[IO]] = ResourceSuiteLocalFixture(
+    "SecretsManagerAlg",
+    Ref.in[Resource[IO, *], IO, Map[SecretId, String]](Map.empty).map(new FakeSecretsManagerAlg(_))
+  )
+
+  private implicit val prettyRequestResource: Resource[IO, CloudFormationCustomResourceRequest[DatabaseMetadata]] => Pretty = {
+    case Resource.Pure(req) => Pretty(_ => req.asJson.spaces2)
+    case other => Pretty { _ =>
+      other.use { req =>
+        req
+          .ResourceProperties
+          .secretIds
+          .traverse(secretsManagerAlg().getSecret(_).flatMap(parser.parse(_).liftTo[IO]))
+          .map { secrets =>
+            req.applyLens(requestTypeLens).set(CloudFormationRequestType.OtherRequestType("template"))
+              .asJson
+              .mapObject(_.add("Materialized Secrets (this key is synthetic and not part of the AWS protocol)", secrets.asJson))
+              .spaces2
+          }
+      }.unsafeRunSync() // this would normally be illegal, but it's probably fine here since this only happens when a test failed anyway
+    }
+  }
+
+  override def munitFixtures = List(secretsManagerAlg)
+
+  private def requestTypeLens[T]: Lens[CloudFormationCustomResourceRequest[T], CloudFormationRequestType] = GenLens[CloudFormationCustomResourceRequest[T]](_.RequestType)
+
+  test("Handler can create and destroy a database with users") {
+    implicit val arbDatabaseMetadata: Arbitrary[Resource[IO, DatabaseMetadata]] = Arbitrary(genDatabaseMetadata[IO, UserConnectionInfo](secretsManagerAlg(), genUserConnectionInfo))
+    implicit val arbReq: Arbitrary[Resource[IO, CloudFormationCustomResourceRequest[DatabaseMetadata]]] = Arbitrary(genWrappedCloudFormationCustomResourceRequest[Resource[IO, *], DatabaseMetadata])
+
+    PropF.forAllF { input: Resource[IO, CloudFormationCustomResourceRequest[DatabaseMetadata]] =>
+      val r = for {
+        requestCapture <- Ref.in[Resource[IO, *], IO, CapturedRequests[IO]](List.empty)
+        client = FakeS3(requestCapture)
+        requestTemplate <- input
+        handler <- new MySqlDatabaseInitHandlerF[IO] {
+          override protected def httpClient: Resource[IO, Client[IO]] = client.pure[Resource[IO, *]]
+          override protected def secretsManagerResource(implicit L: Logger[IO]): Resource[IO, SecretsManagerAlg[Kleisli[IO, Span[IO], *]]] =
+            secretsManagerAlg().mapK(Kleisli.liftK[IO, Span[IO]]).pure[Resource[IO, *]]
+        }.handler
+        secrets <- Resource.eval(requestTemplate.ResourceProperties.secretIds.traverse(secretsManagerAlg().getSecretAs[UserConnectionInfo]))
+      } yield (requestCapture, requestTemplate.applyLens(requestTypeLens).set(_), handler, secrets)
+
+      r.use { case (requestCapture, requestTemplate, handler, secrets) =>
```

Comment:
I vaguely remember you mentioning there exists some danger in calling `.use` before the last possible moment--though I am not confident in remembering exactly what the danger was. Would this issue be a problem here?

Reply:
In this case, this is the last possible moment. We want the resource lifecycle to execute as part of each property test, so that everything gets cleaned up after each random input is executed. Good question!

Reply:
That makes sense--thanks for explaining!